core/src/main/java/org/elasticsearch/cluster/health/ClusterShardHealth.java

```diff
@@ -126,4 +127,38 @@ public void writeTo(final StreamOutput out) throws IOException {
         out.writeVInt(unassignedShards);
         out.writeBoolean(primaryActive);
     }
+
+    /**
+     * Checks if an inactive primary shard should cause the cluster health to go RED.
+     *
+     * Normally, an inactive primary shard in an index should cause the cluster health to be RED.  However,
+     * there are exceptions where a health status of RED is inappropriate, namely in these scenarios:
+     *   1. Index Creation.  When an index is first created, the primary shards are in the initializing state, so
+     *      there is a small window where the cluster health is RED due to the primaries not being activated yet.
+     *      However, this leads to a false sense that the cluster is in an unhealthy state, when in reality, its
+     *      simply a case of needing to wait for the primaries to initialize.
+     *   2. When a cluster is in the recovery state, and the shard never had any allocation ids assigned to it,
+     *      which indicates the index was created and before allocation of the primary occurred for this shard,
+     *      a cluster restart happened.
+     *
+     * Here, we check for these scenarios and set the cluster health to YELLOW if any are applicable.
+     *
+     * NB: this method should *not* be called on active shards nor on non-primary shards.
+     */
+    public static ClusterHealthStatus getInactivePrimaryHealth(final ShardRouting shardRouting, final IndexMetaData indexMetaData) {
+        assert shardRouting.primary();
```

Comment:
add some text to the assertion together with the shardRouting that made it fail.